framework/src/outputs/Exodus.C

```diff
@@ -533,19 +539,23 @@ Exodus::clear()
 void
 Exodus::handleExodusIOMeshRenumbering()
 {
-  // We know exodus_io renumbered on the first write_timestep()
-  if (!_exodus_initialized && !_mesh_contiguous_numbering)
+  // We renumbered our mesh, so we need the other mesh to do the same
+  if (auto * const disp_problem = _problem_ptr->getDisplacedProblem().get(); disp_problem)
   {
-    // We renumbered our mesh, so we need to allow the other mesh to do the same
-    if (auto * const disp_problem = _problem_ptr->getDisplacedProblem().get(); disp_problem)
-    {
-      auto & disp_eq = disp_problem->es();
-      auto & other_mesh = &disp_eq == _es_ptr ? _problem_ptr->mesh().getMesh() : disp_eq.get_mesh();
-      other_mesh.allow_renumbering(true);
-    }
-
-    // Objects that depend on element/node ids are no longer valid
-    _problem_ptr->meshChanged();
-    _mesh_contiguous_numbering = true;
+    auto & disp_eq = disp_problem->es();
+    auto & other_mesh = &disp_eq == _es_ptr ? _problem_ptr->mesh().getMesh() : disp_eq.get_mesh();
+    mooseAssert(
+        !other_mesh.allow_renumbering(),
+        "The only way we shouldn't have contiguous numbering is if we've disabled renumbering");
+    other_mesh.allow_renumbering(true);
+    other_mesh.renumber_nodes_and_elements();
+    // Copying over the comment in MeshOutput::write_equation_systems
+    // Not sure what good going back to false will do here, the
+    // renumbering horses have already left the barn...
+    other_mesh.allow_renumbering(false);
```

Comment:
if I see a mesh with allow_renumbering = false I kind of expect it did not get renumbered

Reply:
This is what is done in the `MeshOutput::write` code. I think the overriding priority is to keep the two meshes in sync

Reply:
And the value of this boolean does in fact change the outcome of calls to `ReplicatedMesh::renumber_nodes_and_elements`

Reply:
You can make an issue/PR to change the libmesh behavior if you wish but it will likely lead to exodiffs

Reply:
The bottom line is that as it stands now you shouldn't see `allow_renumbering(false)` and trust that means the mesh has not or will not be renumbered. `contract` for instance, which is always called post-adaptivity, always calls `renumber_nodes_and_elements` which always renumbers regardless of `_skip_renumber_nodes_and_elements`

Reply:
> I think the overriding priority is to keep the two meshes in sync

Agreed
Good to know on the other items. There are two discussions posts open complaining about renumbering btw

Reply:
Associated libmesh issues: https://github.com/libMesh/libmesh/issues/3966, https://github.com/libMesh/libmesh/issues/1216, and https://github.com/libMesh/libmesh/issues/938